File: spcon/public/js/sales_order.js
// frappe.ui.form.on("Sales Order Item", {
// 	custom_check_rate(frm, cdt, cdn) {
// 		const row = locals[cdt][cdn];

// 		if (!frm.doc.customer) {
// 			frappe.msgprint(__("Please select a customer first."));
// 			return;
// 		}

// 		if (!row.item_code) {
// 			frappe.msgprint(__("Please select an item first."));
// 			return;
// 		}

// 		frappe.call({
// 			method: "spcon.public.py.sales_order.get_last_5_sale_rates",
// 			args: { 
// 				customer: frm.doc.customer,
// 				item_code: row.item_code,
// 			},
// 			freeze: true,
// 			callback: function (r) {
// 				const rates = r.message || [];
// 				show_rate_history_dialog(frm.doc.customer, row.item_code, rates);
// 			},
// 		});
// 	},
// });

// function show_rate_history_dialog(customer, item_code, rates) {
// 	const dialog = new frappe.ui.Dialog({
// 		title: __("Last 5 Invoice Rates"),
// 		fields: [
// 			{
// 				fieldtype: "HTML",
// 				fieldname: "rate_history_html",
// 			},
// 		],
// 		size: "large",
// 	});

// 	const html = rates.length
// 		? `
// 			<div class="table-responsive">
// 				<table class="table table-bordered table-striped">
// 					<thead>
// 						<tr>
// 							<th>${__("Item")}</th>
// 							<th>${__("Rate")}</th>
// 						</tr>
// 					</thead>
// 					<tbody> 
// 						${rates
// 							.map(
// 								(row) => `
// 									<tr>
// 										<td>${frappe.utils.escape_html(row.item_name || row.item_code || "")}</td>
// 										<td style="text-align: right;">${frappe.format(row.rate, { fieldtype: "Currency" }) || ""}</td>
// 									</tr>
// 								`
// 							)
// 							.join("")}
// 					</tbody>
// 				</table>
// 			</div>
// 		`
// 		: `
// 			<p>${__("No submitted sales invoice rates were found for this customer and item.")}</p>
// 		`;

// 	dialog.fields_dict.rate_history_html.$wrapper.html(html);
// 	dialog.show();
// }


frappe.ui.form.on('Sales Order', {
	async cost_center(frm) {
		if (!frm.doc.cost_center) {
			frm.set_value('company_address', '');
			return;
		}

		const r = await frappe.db.get_value('Cost Center', frm.doc.cost_center, 'custom_address');
		const address = r.message?.custom_address || '';
		frm.set_value('company_address', address);
		console.log(address);
		apply_warehouse_filter(frm);
	},
    onload: function(frm) {
        apply_warehouse_filter(frm);
    }
});

function apply_warehouse_filter(frm) {
    if (!frm.doc.cost_center) return;

    frm.set_query("set_warehouse", function() {
        return {
            query: "spcon.public.py.warehouse.get_warehouses_for_cost_center",
            filters: {
                cost_center: frm.doc.cost_center
            }
        };
    });
}




frappe.ui.form.on('Sales Order Item', {
    custom_check_rate: function(frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (!frm.doc.customer || !row.item_code) {
            frappe.msgprint("Customer and Item required");
            return;
        }

        frappe.call({
            method: "spcon.public.py.sales_order.get_latest_rate",
            args: {
                customer: frm.doc.customer,
                item_code: row.item_code
            },
            callback: function(r) {
                if (r.message && r.message.length) {

                    let msg = "<b>Last 5 Rates:</b><br><br>";

                    r.message.forEach(d => {
                        msg += `Invoice: ${d.invoice} | Date: ${d.posting_date} | Rate: ${d.rate}<br>`;
                    });

                    frappe.msgprint(msg);

                } else {
                    frappe.msgprint("No previous Sales Invoice found");
                }
            }
        });
    }
});

File: spcon/public/py/sales_order.py
# import frappe


# @frappe.whitelist()
# def get_last_5_sale_rates(customer, item_code):
# 	if not customer or not item_code:
# 		return []

# 	return frappe.db.sql(
# 		"""
# 		SELECT 
# 			sii.item_code,
# 			sii.item_name,
# 			sii.rate
# 		FROM `tabSales Invoice Item` sii
# 		INNER JOIN `tabSales Invoice` si
# 			ON si.name = sii.parent
# 		WHERE si.docstatus = 1
# 			AND si.customer = %(customer)s
# 			AND sii.item_code = %(item_code)s
# 		ORDER BY si.posting_date DESC, si.modified DESC, sii.idx DESC
# 		LIMIT 5
# 		""",
# 		{"customer": customer, "item_code": item_code},
# 		as_dict=True,
# 	)


import frappe

@frappe.whitelist()
def get_latest_rate(customer, item_code):
    result = frappe.db.sql("""
        SELECT 
            si.name as invoice,
            si.posting_date,
            sii.rate
        FROM `tabSales Invoice Item` sii
        INNER JOIN `tabSales Invoice` si
            ON sii.parent = si.name
        WHERE 
            si.customer = %s
            AND sii.item_code = %s
            AND si.docstatus = 1
        ORDER BY si.posting_date DESC, si.creation DESC
        LIMIT 5
    """, (customer, item_code), as_dict=True)

    return result


# @frappe.whitelist()
# def set_minimum_qty(doc, method=None):
# 	if doc.items:
# 		for item in doc.items:
# 			min_qty = frappe.get_value("Item", item.item_code, "custom_minimum_sale_qty")
# 			if item.qty < min_qty:
# 				frappe.throw(f"Minimum quantity for item {item.item_code} is {min_qty}. Please adjust the quantity accordingly.")	

@frappe.whitelist()
def set_minimum_qty(doc, method=None):
    errors = []

    if doc.items:
        for item in doc.items:
            min_qty = frappe.get_value("Item", item.item_code, "custom_minimum_sale_qty") or 0

            if item.qty < min_qty:
                errors.append(
                    f"Row {item.idx}: Item {item.item_code} → Minimum Qty = {min_qty}, Entered Qty = {item.qty}"
                )

    if errors:
        # frappe.throw("<br>".join(errors))
        frappe.msgprint("<br>".join(errors))